core/src/test/java/io/hyperfoil/core/impl/statistics/StatisticsStoreTest.java

```diff
@@ -0,0 +1,108 @@
+
+import com.fasterxml.jackson.core.JsonFactory;
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.core.io.JsonStringEncoder;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import io.hyperfoil.api.config.*;
+import io.hyperfoil.api.http.HttpMethod;
+import io.hyperfoil.api.statistics.StatisticsSnapshot;
+import io.hyperfoil.core.builders.StepCatalog;
+import io.hyperfoil.core.steps.HttpRequestStep;
+import io.vertx.ext.unit.junit.VertxUnitRunner;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.*;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import static org.junit.Assert.*;
+
+@RunWith(VertxUnitRunner.class)
+public class StatisticsStoreTest {
+
+   @Test
+   public void allJson(){
+      ErgonomicsBuilder ergonomicsBuilder = new ErgonomicsBuilder();
+      ergonomicsBuilder.repeatCookies(true).userAgentFromSession(true);
+      Benchmark benchmark = new BenchmarkBuilder("originalSource", BenchmarkData.EMPTY)
+         .name("benchmarkName")
+         .http().host("localhost").endHttp()
+         .addPhase("phaseName/001/test").always(1)
+         .duration(60_000)
+         .scenario().initialSequence("test")
+         .step(StepCatalog.class).httpRequest(HttpMethod.GET)
+         .sla().addItem().meanResponseTime(1, TimeUnit.MILLISECONDS).endSLA().endList()
+         .endStep()
+         .endSequence().endScenario()
+         .endPhase()
+         .build();
+      Phase phase = benchmark.phases().stream().findAny().get();
+      HttpRequestStep step = (HttpRequestStep) Stream.of(phase.scenario().sequences())
+         .flatMap(s -> Stream.of(s.steps()))
+         .filter(HttpRequestStep.class::isInstance)
+         .findAny().get();
+
+      StatisticsStore store = new StatisticsStore(benchmark,failure-> {});
+
+      StatisticsSnapshot snapshot = new StatisticsSnapshot();
+      snapshot.sequenceId=101;
+      snapshot.requestCount=6;
+      snapshot.histogram.setStartTimeStamp(1*60_000);
+      snapshot.histogram.setEndTimeStamp(2*60_000);
+      snapshot.histogram.recordValue(1_000);
+      snapshot.histogram.recordValue(10_000);
+      snapshot.histogram.recordValue(30_000);
+      snapshot.histogram.recordValue(60_000);
+      snapshot.histogram.recordValue(120_000);
+      snapshot.histogram.recordValue(666_000_000);
+
+      store.record("address",phase.id(), step.id(), "metric1",snapshot);
+      store.recordSessionStats("address",1*60_000,"phaseName/001/test",1,10);
+//      snapshot.sequenceId=101;
+//      snapshot.histogram.setStartTimeStamp(2*60_000);
+//      snapshot.histogram.setEndTimeStamp(3*60_000);
+//      store.record("address",1,1,"metric1",snapshot);
+      snapshot.sequenceId=201;
```

Comment:
I still don't think it's a good idea to overwrite the snapshot object - the one you've recorded on line 66 is probably stored somewhere and you're modifying it (no defensive copy made).

Reply:
You're right, snapshots were not being copied in the calls to mergeSnapshots(...) so I updated it to create a unique snapshot.